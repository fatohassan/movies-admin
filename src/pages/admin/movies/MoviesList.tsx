import axios from "axios";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import LoadingComponent from "../../../components/LoadingComp";

type Movie = {
  id: number;
  original_title: string;
  vote_average: number;
  release_date: string;
  backdrop_path: string;
};

export default function MoviesList() {
  const [movies, setMovies] = useState<Movie[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  const fetchMovie = async () => {
    const res = await axios.get("http://localhost:5000/");
    const data = await res.data;
    setMovies(data);
  };

  useEffect(() => {
    fetchMovie();
    setTimeout(() => {
      setIsLoading(false);
    }, 2000);
  }, []);

  return (
    <div className="container my-4">
      <h2 className="text-center mb-4">Movies</h2>

      <div className="row mb-3">
        <div className="col">
          <Link
            className="btn btn-primary me-1"
            to="/admin/movies/create"
            role="button"
          >
            Create Movie
          </Link>
          <button
            type="button"
            className="btn btn-outline-primary"
            onClick={fetchMovie}
          >
            Refresh
          </button>
        </div>
        <div className="col"></div>
      </div>

      {isLoading ? (
        <LoadingComponent />
      ) : (
        <table className="table">
          <thead>
            <tr>
              <th>ID</th>
              <th>Title</th>
              <th>Rating</th>
              <th>Image</th>
              <th>Release Date</th>
              <th>Action</th>
            </tr>
          </thead>
          <tbody>
            {movies.map((movie, index) => {
              return (
                <tr key={index}>
                  <td>{movie.id}</td>
                  <td>{movie.original_title}</td>
                  <td>{movie.vote_average}</td>
                  <td>
                    <img
                      src={`https://image.tmdb.org/t/p/w1280${movie.backdrop_path}`}
                      width="100"
                      alt="..."
                    />
                  </td>
                  <td>{movie.release_date}</td>
                  <td style={{ width: "10px", whiteSpace: "nowrap" }}>
                    <Link
                      className="btn btn-primary btn-sm me-1"
                      // add the id to be edited
                      to={"/admin/movies/edit/"}
                    >
                      Edit
                    </Link>
                    <button type="button" className="btn btn-danger btn-sm">
                      Delete
                    </button>
                  </td>
                </tr>
              );
            })}
          </tbody>
        </table>
      )}
    </div>
  );
}
