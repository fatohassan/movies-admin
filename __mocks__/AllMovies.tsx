import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { renderHook } from "@testing-library/react";
import React from "react";
import  MoviesList  from "../src/pages/admin/movies/MoviesList";
import { LOAD_MOVIES } from "../src/pages/admin/movieQueries/moviesQueries";

export const AllMovies = {
    getMoviesList: [
    {
      id: "66715c4f88c42067190e070f",
      original_title: "Inside Out 2",
      release_date: "2024-06-11",
      vote_average: "7.761",
      backdrop_path: "/coATv42PoiLqAFKStJiMZs2r6Zb.jpg",
      image: null,
      overview:
        "Teenager Riley's mind headquarters is undergoing a sudden demolition to make room for something entirely unexpected: new Emotions! Joy, Sadness, Anger, Fear and Disgust, who’ve long been running a successful operation by all accounts, aren’t sure how to feel when Anxiety shows up. And it looks like she’s not alone.",
    },
  ],
};

// successful mock of getMovieList
export const successfulMoviesMock: MockedResponse[]= [
    {
        request: {
            query: LOAD_MOVIES
        },
        result: {
            data: AllMovies
        }
    }
] 

export function getAllMoviesWrapper(mockData: MockedResponse[] = []) {
    const wrapper =({children}: React.PropsWithChildren) => (
        <MockedProvider mocks={mockData} addTypename={false}>
            {children}
        </MockedProvider>
    );
    const {result}:any = renderHook(() => MoviesList(), {wrapper})
    return {
        result
    }
}
