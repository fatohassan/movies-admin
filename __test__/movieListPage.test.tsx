import { cleanup, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import MoviesList from "../src/pages/admin/movies/MoviesList";
import React from "react";
import { MockedProvider } from "@apollo/client/testing";
import { listMoviesMock } from "../__mocks__/AllMovies";
import { BrowserRouter } from "react-router-dom";

describe("MoviesList page rendering/navigating", () => {
  afterEach(() => {
    cleanup();
  });

  it("should display contents", async () => {
    render(
      <BrowserRouter>
        <MockedProvider mocks={listMoviesMock} addTypename={false}>
          <MoviesList />
        </MockedProvider>
      </BrowserRouter>
    );
    expect(screen.getByText("Movies")).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /Movies/i })
    ).toBeInTheDocument();
    expect(screen.getByText("Create Movie")).toBeInTheDocument();
    expect(screen.getByText("Refresh")).toBeInTheDocument();
    expect(screen.getByRole("heading")).toHaveTextContent("Movies");
  });

  it("should display table contents", async () => {
    render(
      <BrowserRouter>
        <MockedProvider mocks={listMoviesMock} addTypename={false}>
          <MoviesList />
        </MockedProvider>
      </BrowserRouter>
    );
    expect(await screen.findByRole("img")).toBeInTheDocument();
    expect(await screen.findByText("Inside Out 2")).toBeInTheDocument();
    expect(screen.getByText("Title")).toBeInTheDocument();
    expect(screen.getByText("Rating")).toBeInTheDocument();
    expect(screen.getByText("Image")).toBeInTheDocument();
    expect(screen.getByText("Release Date")).toBeInTheDocument();
    expect(screen.getByText("Action")).toBeInTheDocument();
    expect(await screen.findByText("Edit")).toBeInTheDocument();
    expect(await screen.findByText("Delete")).toBeInTheDocument();
  });

  it("should display routes/links/button", async () => {
    render(
      <BrowserRouter>
        <MockedProvider mocks={listMoviesMock} addTypename={false}>
          <MoviesList />
        </MockedProvider>
      </BrowserRouter>
    );
    await screen.findAllByRole("link");
    screen.getAllByRole("button");
    expect(
      screen.getByRole("button", { name: /Create Movie/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /Refresh/i })
    ).toBeInTheDocument();
    expect(
      await screen.findByRole("button", { name: /Delete/i })
    ).toBeInTheDocument();
    expect(
      await screen.findByRole("link", { name: /Edit/i })
    ).toBeInTheDocument();
  });
});
