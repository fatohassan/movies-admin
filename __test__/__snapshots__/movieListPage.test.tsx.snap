// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList page rendering/navigating should render page correctly 1`] = `
<div
  className="container my-4"
>
  <h2
    className="text-center mb-4"
  >
    Movies
  </h2>
  <div
    className="row mb-3"
  >
    <div
      className="col"
    >
      <a
        className="btn btn-primary me-1"
        href="/admin/movies/create"
        onClick={[Function]}
        role="button"
      >
        Create Movie
      </a>
      <button
        className="btn btn-outline-primary"
        onClick={[Function]}
        type="button"
      >
        Refresh
      </button>
    </div>
    <div
      className="col"
    />
  </div>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <div
      dangerouslySetInnerHTML={
        {
          "__html": "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32">
  <path opacity=".25" d="M16 0 A16 16 0 0 0 16 32 A16 16 0 0 0 16 0 M16 4 A12 12 0 0 1 16 28 A12 12 0 0 1 16 4"/>
  <path d="M16 0 A16 16 0 0 1 32 16 L28 16 A12 12 0 0 0 16 4z">
    <animateTransform attributeName="transform" type="rotate" from="0 16 16" to="360 16 16" dur="0.8s" repeatCount="indefinite" />
  </path>
</svg>
",
        }
      }
      style={
        {
          "fill": "#1d81e2",
          "height": 64,
          "width": 64,
        }
      }
    />
  </div>
</div>
`;
